File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # The problem constraints guarantee to have a head
        if head.next is None:
            return head
        
        tortoise = hare = head
        
        while hare and hare.next:
            hare = hare.next.next
            tortoise = tortoise.next
        
        return tortoise

File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// Input: head = [1,2,3,4,5]
// Output: [3,4,5]

// Input: head = [1,2,3,4,5,6]
// Output: [4,5,6]

// we can have a fast pointer
// a slow pointer

// The number of nodes in the list is in the range [1, 100].


class Solution {
public:

    // Input: head = [1,2,3,4,5]
    // Output: [3,4,5]

    // Input: head = [1,2,3,4,5,6]
    // Output: [4,5,6]

    // Input: head = [1]
    // Output: [1]

    // Input: head = [1,2]
    // Output: [2]

    ListNode* middleNode(ListNode* head) {

        ListNode* fast = head;
        ListNode* slow = head;

        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        return slow;
    }
};